docsite/rst/developing_program_flow_modules.rst

```diff
@@ -0,0 +1,392 @@
+``ANSIBLE_VERSION``, ``SELINUX_SPECIAL_FS`` and ``SYSLOG_FACILITY``.  In
+:ref:`module_replacer`, ``ANSIBLE_VERSION`` and ``SELINUX_SPECIAL_FS`` were
+substituted into the global variables
+:code:`ansible.module_utils.basic.ANSIBLE_VERSION` and
+:code:`ansible.module_utils.basic.SELINUX_SPECIAL_FS`.  ``SYSLOG_FACILITY`` didn't
+get placed into a variable.  Instead, any occurrences of the string
+``syslog.LOG_USER`` in the combined module file were replaced with ``syslog.``
+followed by the string contained in ``SYSLOG_FACILITY``.  All of these have
+changed in :ref:`ziploader`.
+
+The ansible verison can now be used by a module by importing ``__version__``
+from ansible::
+
+    from ansible import __version__
+    module.exit_json({'msg': 'module invoked by ansible %s' % __version__})
+
+For now :code:`ANSIBLE_VERSION` is also available at its old location inside of
+``ansible.module_utils.basic`` but that will eventually be removed.
+
+``SELINUX_SPECIAL_FS`` and  ``SYSLOG_FACILITY`` have changed much more.
+:ref:`ziploader` passes these as another json-ified string inside of the
+:envvar:`ANSIBLE_MODULE_CONSTANTS` environment variable.  When
+``ansible.module_utils.basic`` is imported it places this string in the global
+variable :code:`ansible.module_utils.basic.MODULE_CONSTANTS` and removes it from
+the environment.  The constants are parsed when an :class:`AnsibleModule` is
+instantiated.  Modules shouldn't access any of those directly.  Instead they
+should instantiate an :class:`AnsibleModule` and use
+:attr:`AnsibleModule.constants` to access the parsed version of these values.
+
+Unlike the ``ANSIBLE_ARGS`` and ``ANSIBLE_VERSION`` where some efforts were
+made to keep the old backwards compatible globals available, these two
+constants are not available at their old names.  This is a combination of the
+degree to which these are internal to the needs of ``module_utils.basic`` and
+(in the case of ``SYSLOG_FACILITY``) how hacky and unsafe the previous
+implementation was.
+
+Porting code from the :ref:`module_replacer` method of getting
+``SYSLOG_FACILITY`` to the new one is a little more tricky than the other
+constants and args due to just how hacky the old way was.  Here's an example
+of using it in the new way::
+
+        import syslog
+        facility_name = module.constants.get('SYSLOG_FACILITY')
+        facility = getattr(syslog, facility_name)
+        syslog.openlog(str(module), 0, facility)
+
+.. _flow_special_considerations:
+
+Special Considerations
+----------------------
+
+.. _flow_pipelining:
+
+Pipelining
+^^^^^^^^^^
+
+Ansible can transfer a module to a remote machine in two ways.  Either it can
+write out the module to a temporary file on the remote host and then use
+a second connection to the remote host to execute it with the interpreter that
+the module needs or it can use what's known as pipelining to execute the module
+by piping it into the remote interpreter's stdin.  Pipelining only works with
```

Comment:
Maybe make the "Either" sentence into two bullet points, since it's a big one with lots of good info in it?

Ansible can transfer a module to a remote machine in one of two ways::  
* it can write out the module to a temporary file on the remote host and then use a second connection to the remote host to execute it with the interpreter that the module needs
* or it can use what's known as pipelining to execute the module by piping it into the remote interpreter's stdin.